### PR TITLE
Paper: multinterp: A Unified Interface for Multivariate Interpolation in the Scientific Python Ecosystem

### PR DESCRIPTION
**Pull request by alanlujan91**
If you are creating this PR in order to submit a draft of your paper, please name your PR with `Paper: <title>`. An editor will then add a `paper` label and GitHub Actions will be run to check and build your paper.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#preview-your-paper) for more information.

_Editor_:  Amey Ambade @ameyxd 

_Reviewers_: 
- Abhay Paroha @aparoha
- Gajendra Deshpande @gcdeshpande
- Jennifer Yoon @JennEYoon

### CODE REVIEW

**Comment by ameyxd** [comment]
@alanlujan91 can you recheck the DOIs that fail checks? As long as they are valid DOIs we should be okay. 

you may be able to add citation keys you want to ignore if their DOIs don't exist in myst.yml under error_rules:

```error_rules:
  - rule: doi-exists
    severity: ignore
    keys:
      - abc
      - def01

**Comment by ameyxd** [comment]
@alanlujan91 are you skipping valid DOIs? If so, please refrain and try the checks again. 😅

**Comment by alanlujan91** [comment]
@ameyxd I believe the ones I'm skipping don't have DOIs. I can try to look again

**Comment by ameyxd** [comment]
Okay! please check "Paszke2019" in particular. lmk when complete and passing checks.

**Comment by alanlujan91** [comment]
@ameyxd Oh I remember what happened.

I found https://dl.acm.org/doi/10.5555/3454287.3455008
so I put that on my bib, then pushed, but mystmd didn't like the doi, probably because of this
image
so then I set to ignore, but mystmd still didn't like it, so I removed it altogether from the bib
@rowanc1 any advice?

**Comment by ameyxd** [comment]
Review reminders sent to @gcdeshpande and @aparoha

**Comment by cbcunc** [comment]
> so then I set to ignore, but mystmd still didn't like it, so I removed it altogether from the bib @rowanc1 any advice?

@rowanc1 Any comment?

File: papers/alan_lujan/main.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+As we might imagine, interpolation on regular grids is much easier than interpolation on irregular grids as we are able to exploit the structure of the grid to make predictions about the function's behavior between known values. Irregular grid interpolation is much more difficult, and often requires *regularizing* and/or *regression* techniques to make predictions about the function's behavior between known values. `multinterp` aims to provide a comprehensive set of tools for both regular and irregular grid interpolation, and we will discuss some of these tools in the following sections.
+
+```{list-table} Grids and structures implemented in "multinterp".
```
Comment (aparoha) [inline papers/alan_lujan/main.md:45]: Question: Do you have plans to support other interpolations e.g. spline, inverse distance weight, natural neighbor etc.?

File: papers/alan_lujan/main.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+The set of input values on which we know the function's output values is called a **grid**. A grid (or input grid) of values can be represented in many ways, depending on its underlying structure. The broadest categories of grids are regular or structured grids, and irregular or unstructured grids. Regular grids are those where the input values are arranged in a regular pattern, such as a triangle or a quadrangle. Irregular grids are those where the input values are not arranged in a particularly structured way and can seem to be scattered randomly across the input space.
```
Comment (aparoha) [inline papers/alan_lujan/main.md]: The introduction of technical terms and concepts could be more gradual and systematic, ensuring the audience can follow the content smoothly. The 2nd paragraph directly jumps to grid interpolation without setting the context before.

File: papers/alan_lujan/main.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+The `multinterp` project seeks to change this. Its goal is to provide a unified, comprehensive, and flexible framework for multivariate interpolation in Python. This framework will streamline workflows by offering:
+
+- Unified Interface: A consistent API for interpolation, regardless of data structure or desired backend, reducing the learning curve and promoting code reusability.
+- Hardware Adaptability: Seamless support for CPU (NumPy, SciPy), parallel (Numba), and GPU (CuPy, PyTorch, JAX) backends, empowering users to optimize performance based on their computational resources.
+- Broad Functionality: Tools for regular/rectilinear interpolation, multivalued interpolation, and derivative calculations, addressing a wide range of scientific problems.
+
+The multinterp package (<https://github.com/alanlujan91/multinterp>) is currently in its beta stage.  It offers a strong foundation but welcomes community contributions to reach its full potential.  We invite collaboration to improve documentation, expand the test suite, and ensure the codebase aligns with the highest standards of Python package development.
```
Comment (aparoha) [inline papers/alan_lujan/main.md]: While the text mentions documentation and community contributions, specific links or guidelines for how users can contribute or access detailed documentation are not provided. Adding clear references or links to supplementary materials would enhance the text's utility.

**Review by aparoha — CHANGES_REQUESTED** [review]
Thanks for raising your PR. The overall concepts look promising. I have provided in-line comments as well. Overall comments:-
1. The text could benefit from clearer organization and structure. For instance, some sections seem to repeat concepts without introducing new information or advancing the narrative.
2. The introduction of technical terms and concepts could be more gradual and systematic, ensuring the audience can follow the content smoothly.

**Comment by rowanc1** [comment]
Hi @alanlujan91 - we talked about this today in our editorial team meeting and would love to showcase your article with computational capabilities and reproducibility (as [per this note](https://github.com/scipy-conference/scipy_proceedings?tab=readme-ov-file#a-note-on-notebooks-for-2024) in the readme!).

I think following the review comment by @aparoha "some sections seem to repeat concepts without introducing new information or advancing the narrative.", it looks as though you are including the full notebooks, rather than individual cells. There is a way in MyST to include individual cells or outputs that keep links to the original location of the supporting materials. The [documentation is here](https://mystmd.org/guide/reuse-jupyter-outputs). That could be a way to change how you are showcasing the information and better highlight and link to your supporting materials.

If you are open to it I would love to work on you to change the use of MyST slightly to better highlight the computational aspects. Let me know if you have time this week or next week?

**Comment by alanlujan91** [comment]
@rowanc1 that would be great!

I am at a conference this week, so I haven't had time to address the previous comments, but will be available next week to discuss how we can improve this work

**Comment by rowanc1** [comment]
Fantastic - I will reach out to you separately and get a meeting scheduled soon. Very very excited to show off your content in the SciPy Proceedings this year, and elevate the use of Jupyter Notebooks and reproducibility. 🚀

**Comment by JennEYoon** [comment]
Re: https://github.com/alanlujan91/scipy_proceedings/pull/1

Dear Alan @alanlujan91, 
Thank you for the opportunity to review Multinterp paper for SciPy Conference Proceedings 2024.  All of my comments are to improve readability for the general reader.  Please feel free to incorporate or ignore any of my suggestions. 

File edit:   main.md  `[[ suggestions enclosed between double brackets ]]`
Add file:   comments_figs_nbs.md ( suggestions on figures )  
screenshots (hand markup drawings) in `.\figures\ 

Sincerely,  
Jennifer Yoon, @JennEYoon  
best email:  mail@JenniferYoon.com 
July 24, 2024  PR1 submit

**Comment by ameyxd** [comment]
@JennEYoon thanks for your reviews, please leave your commit as comments in this thread. You may refer to the work by other reviewers. Let us know if you have issues!

File: papers/alan_lujan/main.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+Unstructured grids are irregular and often require a triangulation step which might be computationally expensive and time-consuming.
+```
+
+```{include} notebooks/Unstructured_Interpolation.ipynb
+```
+
+## Conclusion
+
+Multivariate interpolation is a cornerstone of scientific computing, yet the Python ecosystem (@Oliphant2007) presents a fragmented landscape of tools. While individually powerful, these packages often lack a unified interface. This fragmentation makes it difficult for researchers to experiment with different interpolation methods, optimize performance across diverse hardware, and handle varying data structures (regular, rectilinear, curvilinear, unstructured).
+
```
Comment (JennEYoon) [inline papers/alan_lujan/main.md:555]: Is there a way to give a nod to PyTorch and TensorFlow for having call-backs, hooks, and function wrappers to allow a user to swap out an optimization function or module mid-stream? They do not cover all the use cases of the `multinterp` package, but some effort went into developing a layered API to cover varying use cases.  

NumPy also has structured data type, that can be used for custom data type and hierarchial data structures.  It can be seen as an attempt to provide a flexible (customizable) user interface, even though its aim and scope is different from 'multinterp.' (NumPy structured datatype's goal seems mostly for C code interface and optimized C module or C numerical recipe interface and explicit memory control or memory layout control.) The general reader may appreciate having some context. This package may be viewed as a further development of previous efforts at a flexible user interface for users of varying data types and data geometries. 
(See structured arrays in https://numpy.org/doc/stable/user/basics.rec.html)

File: papers/alan_lujan/main.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
+---
+# Ensure that this title is the same as the one in `myst.yml`
+title: multinterp
+subtitle: A Unified Interface for Multivariate Interpolation in the Scientific Python Ecosystem
+abstract: |
+  Multivariate interpolation is a fundamental tool in scientific computing, yet the Python ecosystem offers a fragmented landscape of specialized tools. This fragmentation hinders code reusability, experimentation, and efficient deployment across diverse hardware. To address this challenge, I've developed the `multinterp` package. It provides a unified interface for regular/rectilinear interpolation, supports serial (NumPy/SciPy), parallel (Numba), and GPU (CuPy, PyTorch, JAX) backends, and includes tools for multivalued interpolation and interpolation of derivatives.
+exports:
+  - format: pdf
+---
+
```
Comment (JennEYoon) [inline papers/alan_lujan/main.md:8]: Question on abstract: did you mean "regular/irregular interpolation" in your abstract, in place of "regular/rectilinear interpolation"? 
Replace "I've developed the `multinterp` package" to "The `multinterp` package was developed to"

File: papers/alan_lujan/main.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  - Quadrilateral mesh
+* - Unstructured
+  - Irregular
+  - Random 
+```
+
+## Rectilinear Interpolation
+
+A *rectilinear* grid is a regular grid where the input values are arranged in a *rectangular* (in 2D) or *hyper-rectangular* (in higher dimensions) pattern. Moreover, they can be represented by the tensor product of monotonically increasing vectors along each dimension. For example, a 2D rectilinear grid can be represented by two 1D arrays of increasing values, such as $x = [x_0, x_1, x_2, \cdots, x_n]$ and $y = [y_0, y_1, y_2, \cdots, y_m]$, where $x_i > x_j$ and $y_i > y_j$ $\forall i > j$, and the input grid is then represented by $x \times y$ of dimensions $n \times m$. This allows for a very simple and efficient interpolation algorithm, as we can easily find and use the nearest known values to make predictions about the function's behavior in the unknown space.
+
```
Comment (JennEYoon) [inline papers/alan_lujan/main.md:114]: Given your later example, it might be better to make j > i.  
Also vectors are "monotonically increasing," but x_i > x_j and y_i > y_j describe a decreasing function, from left to right.   

In $\forall i > j$, suggest replacing symbol with words "for all i > j," to help non-math readers.

File: papers/alan_lujan/main.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+```{include} notebooks/Curvilinear_Interpolation.ipynb
+```
+
+## Unstructured Interpolation
+
+```{figure} figures/UnstructuredInterpolation
+:label: unstructured
+:alt: Unstructured grids are irregular and often require a triangulation step which might be computationally expensive and time-consuming.
+:align: center
+
```
Comment (JennEYoon) [inline papers/alan_lujan/main.md:388]: Memo, the alt-text sections in code blocks do not wrap properly when viewed on an iPad/Tablet device.  Curvenotes build version works properly. So this issue only comes up when viewed direcly from a GitHub repo.  Not sure if this is an issue for proceedings.  
(Also alt-text wrap issue in Rectilinear Interpolation and Curvilinear Interpolation sections.)

File: papers/alan_lujan/main.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+An additional advantage of `scipy`'s `map_coordinates` is that it has been extended and integrated to various backends, such as `cupy` ([`cupyx.scipy.ndimage.map_coordinates`](https://docs.cupy.dev/en/stable/reference/generated/cupyx.scipy.ndimage.map_coordinates.html)) and `jax` ([`jax.scipy.ndimage.map_coordinates`](https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.ndimage.map_coordinates.html)). This allows for easy and efficient interpolation on GPUs and TPUs, which can be orders of magnitude faster than CPU interpolation. For wider compatibility, `multinterp` also provides a `numba` and `pytorch` implementation of `map_coordinates`, which broadens the range of hardware that can be used for interpolation.
+
+```{include} notebooks/Multivariate_Interpolation.ipynb
+```
+
```
Comment (JennEYoon) [inline papers/alan_lujan/main.md:165]: Multilinear_Interpolation.ipynb comment
```
def squared_coords(x, y):
    return x**2 + y**2
```
It will be helpful to provide a short reason for choosing the squared coordinates function. Example: A squared x and y coordinates function is used to draw a figure whose grid geometry looks like a curved sheet (bowl) in 3D projection. 

This closed-form solution function, for which all points along the curved surface are known, is used as the baseline model. From this known model, we can draw sample points to approximate similarly shaped unknown functions.

Comment (JennEYoon) [inline papers/alan_lujan/main.md:165]: Multivariate_Interpolation.ipynb: figures comment
fig 1: 
  missing title, missing axes labels (x, y, z).  
  suggest adding color shading to increase contrast (center to edges)  
screenshot3

Comment (JennEYoon) [inline papers/alan_lujan/main.md:165]: fig 2: interpolated figures  
   missing title, axes labels (x, y z)  
   suggest ways to show this figure is interpolated.  
Possibly blow up a small area with enhanced pixelation.  
screenshot4

File: papers/alan_lujan/main.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+### Derivatives
+
+The `multinterp` package also allows for the calculation of derivatives of the interpolated function defined on a rectilinear grid. This is done by using the function `get_grad`, which wraps numpy's `gradient` function to calculate the gradient of the interpolated function at the given coordinates.
+
+```{include} notebooks/Multivariate_Interpolation_with_Derivatives.ipynb
+```
+
```
Comment (JennEYoon) [inline papers/alan_lujan/main.md:242]: Multivariate_Interpolation_with_Derivatives.ipynb comment
fig1, fig2, fig3, fig4 
   Axes need labels on all 4 figures (x, y, z) or (dz/dx, dz/dy, f(z)).  

Difficult to see the relationship between first group of 2 figures and second group of 2 figures (partial derivatives). Perhaps some lines can be drawn to show an example of the original function and its partial derivatives.  I leave this decision up to the author.  
screenshot5

File: papers/alan_lujan/main.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+:label: unstructured
+:alt: Unstructured grids are irregular and often require a triangulation step which might be computationally expensive and time-consuming.
+:align: center
+
+Unstructured grids are irregular and often require a triangulation step which might be computationally expensive and time-consuming.
+```
+
+```{include} notebooks/Unstructured_Interpolation.ipynb
+```
+
```
Comment (JennEYoon) [inline papers/alan_lujan/main.md:494]: Unstructured_Interpolation.ipynb comment
For figures after the 1st one, (group 1: nearest, linear, cubic, radial basis) and (group 2: original, gaussian process regression), eye-balling the differences in the figures maybe easier with grid lines drawn in white or black ink.  Also just a suggestion.  

screenshot1

Comment (JennEYoon) [inline papers/alan_lujan/main.md:494]: screenshot2

File: papers/alan_lujan/main.md (hunk 3 above)
Comment (JennEYoon) [inline papers/alan_lujan/main.md]: Curvenoted build version comment  
This paper when viewed on-screen, right-bottom screen area shows a list of related files. If possible, files should be listed in its order of appearance.   

**Supporting Documents**  
*(shown in the order of appearance)*

Multivariate_Interpolation.ipynb
Multivariate_Interpolation_with_Derivatives.ipynb
Multivalued_Interpolation.ipynb
Curvilinear_Interpoliation.ipynb
Unstructured_Interpolation.ipynb
manim_notebook.ipynb (figure animation on Curvenotes build)
figures.ipynb (add?)

**Review by JennEYoon — COMMENTED** [review]
Review comments moved from a PR on author's multinterp branch
   ( https://github.com/alanlujan91/scipy_proceedings/pulls )  
to scipy_proceedings/pull/937 comments section.  
Jennifer 
Sunday August 4, 2024 update

**Comment by ameyxd** [comment]
@JennEYoon, @aparoha - can you confirm the author's updates are satisfactory for your feedback and give me a thumbs up! 🙂

**Comment by JennEYoon** [comment]
Looks good.  All of my comments were up to the author's discretion. :-)

File: papers/alan_lujan/main.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+:label: bilinear
+:alt: A non-uniformly spaced rectilinear grid can be transformed into a uniformly spaced coordinate grid (and vice versa).
+:align: center
+
+A non-uniformly spaced rectilinear grid can be transformed into a uniformly spaced coordinate grid (and vice versa).
+```
+
+### Multilinear Interpolation
+
+`multinterp` provides a simple and efficient implementation of *multilinear interpolation* for various backends (`numpy` (@Harris2020), `scipy` (@Virtanen2020), `numba` (@Lam2015), `cupy` (@Okuta2017), `pytorch` (@Paszke2019), and `jax` (@Bradbury2018)) via its `multinterp` function. From the remaining of this section, `multinterp` refers to the `multinterp` function in `multinterp` package, unless otherwise specified.
```
Comment (JennEYoon) [inline papers/alan_lujan/main.md]: Use "For the remainder of this section..."  instead of "From the remaining of this section"...  Sorry missed moving this comment before; JY

**Review by JennEYoon — COMMENTED** [review]
Added an edit forgot to copy over.

**Comment by stefanv** [comment]
Hi folks, it doesn't look like the performance comparison made it to the published paper?

https://github.com/scipy-conference/scipy_proceedings/pull/937/files#diff-c6654a0adee03d2551feb82286acfd2863716a72e411ad6435c1bc1753445071R503

**Comment by fwkoch** [comment]
Hey @stefanv - thanks for noticing this. As far as I can tell, labels referenced in the `Performance Comparison` section do not exist in any of the notebooks provided (`fig:performance_comparison_2d`, `fig:performance_comparison_3d`, `fig:backend_comparison_2d`, and `fig:backend_comparison_3d`).

If the notebooks with these figures are added, we can update the proceedings site and PDF! cc @alanlujan91

**Comment by alanlujan91** [comment]
thanks for catching that! I will make a PR against `2024`
